README.MD
# SimpleAuthService

SimpleAuthService is designed to be a lightweight user management component for microservice architectures.

SimpleAuthService provides a [REST API](https://bluberrielabs.github.io/SimpleAuthService/) to allow users to create and log into accounts, exchanging login credentials (username/password or Facebook tokens) for [JSON Web Tokens](https://jwt.io).

User passwords are hashed using PBKDF2 with SHA512 before storage.

## Deployment

SimpleAuthService requires Tomcat (or similar servlet engine) and optionally a database. There are several ways to deploy the software. 

### Docker

The easiest way to run the software is using the preconfigured docker image. 

`docker pull bluberrieio/simpleauthservice:latest`

Once downloaded, configure the sas.properties file (see next section) and JWT signing certificates and place them in a folder that is mounted at /etc/sas in the docker image. The Docker container used a MySQL database backend, the configuration details of which are given in the [example sas.properties file](https://github.com/BluberrieLabs/SimpleAuthService/blob/master/docs/sas.\
properties.example), and which should not be changed. 

For instance, if you place the files in /home/user/sas, you could run SimpleAuthService with the command: 

`docker run -v /home/user/sas:/etc/sas -p 127.0.0.1:8080:8080 -i -t bluberrieio/simpleauthservice`

You can then get the openapi.json specification document to check the service is running with:

`curl http://localhost:8080/sas/v1/swagger.json`

### Binary

To deploy the war file to the Java servlet container of your choice. By default, it will use an on disk h2 database. If you would like to use a different RDBMS such as MySQL, create an empty database and user and configure the details in `sas.properties`. See the next section for more details. 

### Building from source

You can build the war file yourself from the latest sourcecode using maven with `mvn compile war:war`.

## Configuration

The service is configured via a Java properties file named `sas.properties`, read from one of the following locations:

* `/etc/sas/sas.properties`
* `$HOME/.sas/sas.properties`

An example sas.properties file can be found in the [docs](https://github.com/BluberrieLabs/SimpleAuthService/blob/master/docs/sas.properties.example) folder. 

The following properties should be set to configured the service:

* `redirecturl` - the URL email confirmations are redirected to (default value="https://www.google.com")
* `sas-user` - the preset admin username (default value="admin@foo.com")
* `sas-passwd` - the preset admin password (default value="changeme")
* `apikey` - key used to secure API operations (default value="1234567")
* `jwt-audience` - the audience value contained in issued JWT tokens (default value="blahblah")
* `jwt-lifetime` - lifetime of issued JWD tokens in miliseconds (default value="86400000")
* `jwt-web-lifetime` - lifetime of JWT tokens issued for web logins in miliseconds (default value="43200000")
* `pwd-reset-lifetime` - lifetime of password reset tokens in miliseconds (default value="3600000")
* `jwt-refresh-lifetime` - lifetime of JWT refresh token (default value="31536000000")
* `jwt-refresh-audience` - the audience value contained in issued JWT refresh tokens (default value="sas-renewer")
* `jwt-issuer` - the issuer value contained in issued JWT tokens (default value="foo.bar.com")
* `facebook-id` - Facebook app ID for Facebook login
* `facebook-secret` - Facebook app secret for Facebook login
* `publickey` - path to public key used to sign JWT tokens
* `privatekey` - path to private key used to sign JWT tokens
* `serverurl` - URL of running service (e.g. https://localhost/sas/v1)
* `reseturl` - the URL of the page to visit to reset a forgotten password.

### Email

You can configure SimpleAuthService to send emails to send automated emails to users for the following:

1. To confirm their email address. When they click on the confirmation link, they are redirected to `redirecturl`.
2. When the address is confirmed, a welcome email is sent.
3. When the user requests to reset their password, a reset email is sent with a link to `reseturl`.

To configure email sending, the following SMTP server details must be set:

* `email-user` - username used to log on to SMTP server
* `email-password` - SMTP server password
* `mail.smtp.host` - SMTP server hostname
* `mail.smtp.starttls.enable` - SMTP server should use TLS (boolean)
* `mail.smtp.auth` - SMTP server requires auth (boolean)

### Database

By default, SimpleAuthService maintains the state of the service to a h2 database file on disk.

SimpleAuthService users [Hibernate](http://hibernate.org/orm/) to map persistent objects to database tables. You can change the backend database used by configuring the following hibernate properties in the sas.properties file:

* `hibernate.connection.driver_class` - database driver to use (default value="org.h2.Driver")
* `hibernate.connection.url` - URL of database to use (default value="jdbc:h2:file:/tmp/sas;DB_CLOSE_DELAY=-1;MVCC=TRU")
* `hibernate.connection.username` - database username (default value="sa")
* `hibernate.connection.password` - database password
* `hibernate.dialect` - SQL dialect to use (default value="org.hibernate.dialect.H2Dialect")

Adding other `hibernated.*` properties to sas.properties should allow you to override standard hibernate.cfg.xml properties. See the Hibernate documentation for more details.

### Certificates

To generate signed JWT tokens, SimpleAuthService requires the access to the public and private components of a digital certificate.

## API

A Swagger description of the API is available from the service. For example, if the service is deployed on localhost port 80, the Swagger document is available at http://localhost/sas/v1/swagger.json.

A full description of the API is available [here](https://bluberrielabs.github.io/SimpleAuthService/).

## Operational Notes

A login provides two credentials: a short-lived user JWT token and a longer lived renewal token. The renewal token can be used to renew the user token, without prompting the user for their login credentials.

The renewal token should be stored securely, in a mobile app keystore for example. When used for web logins, the renewal token is not supplied.

It is advisable to deploy the [SAS-UI](https://github.com/BluberrieLabs/SAS-UI) component to configure the service and manage users.

SimpleAuthService supports two user roles:
1. USER - for everyday users
2. ADMINISTRATOR - required to access the admin services.

The initial user account (configured in the sas.properties file) is given the ADMINISTRATOR role, and allowed to grant that role to other users.

## Limitations

Currently, the service can only generate tokens for a single JWT audience.

The tool is designed to be very lightweight and provide the minimum functionality required to run a mobile app or website where the backend is secured with JWT tokens. It does not implement a full OAuth or OpenID flow.

## Licence

(C) Copyright 2017-2018 Bluberrie Labs (http://bluberrie.io/).

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this software except in compliance with the License. You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.

See the License for the specific language governing permissions and limitations under the License.


Changelog

SimpleAuthService v1.0.0 - Fri 19 Jan 16:18:27 GMT 2018
Updated readme
added docs/api spec
merge
Init repo
Initial commit


src/main/java/io/bluberrie/auth/simplauthservice/Configuration.java
/*
 * (C) Copyright 2017 Bluberrie Labs (http://bluberrie.io/).
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
package io.bluberrie.auth.simplauthservice;

import java.io.File;
import java.io.FileReader;
import java.security.Key;
import java.security.KeyFactory;
import java.security.Security;
import java.sql.Driver;
import java.sql.DriverManager;
import java.sql.SQLException;
import java.util.Collection;
import java.util.HashMap;
import java.util.Map;
import java.util.Properties;
import java.util.Set;

import javax.script.ScriptEngine;
import javax.script.ScriptEngineManager;
import javax.servlet.ServletContext;

import org.bouncycastle.jce.provider.BouncyCastleProvider;
import org.hibernate.Session;
import org.hibernate.SessionFactory;
import org.hibernate.boot.registry.StandardServiceRegistry;
import org.hibernate.boot.registry.StandardServiceRegistryBuilder;
import org.hibernate.internal.SessionFactoryImpl;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import io.bluberrie.auth.simplauthservice.mail.MailSender;


public class Configuration implements Map {

	private SessionFactory sessionFactory;
	private final StandardServiceRegistry registry;

	private Key privateKey;
	private Key publicKey;

	private MailSender mailer;

	private static final Logger LOG = LoggerFactory
			.getLogger(Configuration.class);

	private static final HashMap<String, Configuration> instances = new HashMap<>();
	private static final ScriptEngine groovy = new ScriptEngineManager().getEngineByName("groovy");

	//default locations for the configuration file
	public static File[] locations = {
			new File("/etc/sas/sas.properties"),
			new File(new File(System.getProperty("user.home")), ".sas/sas.properties")
	};



	public static synchronized Configuration getInstance(String context) {
		if (!instances.containsKey(context)) {
			instances.put(context, new Configuration(context));
		}
		return instances.get(context);
	}

	public static synchronized Configuration forContext(ServletContext ctx) {
		String ctxpath=ctx.getContextPath();

		if (ctxpath.length() > 1) {
			return getInstance(ctxpath.substring(1));
		} else {
			return getInstance();
		}
	}

	public static synchronized Configuration getInstance() {
		return Configuration.getInstance("default");
	}

	private final Properties props;
	private final String context;

	public Configuration(String context) {

		this.props=new Properties();

		//default properties
		props.setProperty("redirecturl", "https://www.google.com");
		props.setProperty("sas-user", "admin@foo.com");
		props.setProperty("sas-passwd", "changeme");
		props.setProperty("apikey", "1234567");
		props.setProperty("jwt-audience", "blahblah");
		props.setProperty("jwt-lifetime", "86400000");
		props.setProperty("jwt-web-lifetime", "43200000");
		props.setProperty("pwd-reset-lifetime", "3600000");
		props.setProperty("jwt-refresh-lifetime", "31536000000");
		props.setProperty("jwt-refresh-audience", "sas-renewer");
		props.setProperty("jwt-issuer", "foo.bar.com");
		props.setProperty("facebook-id", "");
		props.setProperty("facebook-secret", "");
		props.setProperty("publickey", "");
		props.setProperty("privatekey", "");

		this.context=context;

		for (File location : locations) {
			if (!location.exists()) {
				continue;
			}
			LOG.info("Reading configuration from " + location);
			try {
				props.load(new FileReader(location));

				LOG.debug(this.toString());
			} catch (Exception ex) {
				LOG.error("Failed to load configuration from " + location, null, ex);
			}
		}
		//open connection      

		//load config defaults. These can be overridden in the sas.properties file
		org.hibernate.cfg.Configuration config = new org.hibernate.cfg.Configuration();
		config.configure("/hibernate.cfg.xml");

		config.addAnnotatedClass(io.bluberrie.auth.simplauthservice.persist.SimpleUser.class);
		config.addAnnotatedClass(io.bluberrie.auth.simplauthservice.persist.SimpleEmail.class);
		config.addAnnotatedClass(io.bluberrie.auth.simplauthservice.persist.SimpleRefreshToken.class);

		//LOG.info("Hibernate config 1: "+config.toString());

		LOG.debug("Hibernate config 2: "+config.getProperties().toString());

		for (String hkey: (Set<String>) this.keySet()) {
			if (hkey.startsWith("hibernate.")) {
				try {
					LOG.info("Adding hibernate key "+hkey + ", value: "+(String)this.get(hkey));
					config.getProperties().remove(hkey);

					config.setProperty(hkey, (String)this.get(hkey));

				} catch (Exception ex) {
					LOG.error("Exception", ex);
				}
			}

		}

		LOG.debug("Hibernate config 3: "+config.getProperties().toString());


		registry = new StandardServiceRegistryBuilder().applySettings(config.getProperties()).build();

		try {
			sessionFactory = config.buildSessionFactory(registry);

			LOG.info("Created hibernate session factory");
		} catch (Exception e) {

			LOG.error("Can't create session factory", e);

			StandardServiceRegistryBuilder.destroy(registry);
		}

		Security.addProvider(new BouncyCastleProvider());
		try {
			KeyFactory factory = KeyFactory.getInstance("RSA", "BC");

			publicKey = Util.loadPublicKey(factory, (String) this.get("publickey"));
			privateKey = Util.loadPrivateKey(factory, (String) this.get("privatekey"));

			//pass properties to the emailer
			Properties mailProps = new Properties();
			for (String hkey: (Set<String>) this.keySet()) {
				if (hkey.startsWith("mail.")) {
					mailProps.setProperty(hkey, (String)this.get(hkey));
				}
			}

			this.mailer = new MailSender(mailProps, this);


		} catch (Exception e) {
			LOG.error("Error loading keys", e);
		}


	}



	public void closeConnection() {
		SessionFactoryImpl sessionFactoryImpl = (SessionFactoryImpl) sessionFactory;
		String url = props.get("hibernate.connection.url").toString();

		LOG.info("CLOSING CONNECTIONS");
		if ( sessionFactory != null ) {
			sessionFactory.close();
		}
		StandardServiceRegistryBuilder.destroy(registry);

		LOG.info("trying to unload driver for "+url);

		Driver driver;
		try {
			driver = DriverManager.getDriver(url);
			DriverManager.deregisterDriver(driver);

		} catch (SQLException e) {
			LOG.error("Couldn't unload driver");
		}


	}


	@Override
	public void finalize() {
		this.closeConnection();
	}

	public Session getSession() {
		return sessionFactory.openSession();
	}

	public MailSender getMailSender() {
		return mailer;
	}

	public Key getJWTPrivateKey() {
		return privateKey;
	}

	public Key getJWTPublicKey() {
		return publicKey;
	}

	public String getPublicKeyFile() {
		return (String) this.get("publickey");
	}

	public String getServerURL() {
		return (String) this.get("serverurl");
	}

	public String getResetURL() {
		return (String) this.get("reseturl");
	}

	public String getDefaultUser() {
		return (String) this.get("sas-user");
	}

	public String getTokenIssuer() {
		return (String) this.get("jwt-issuer");
	}

	public String getConfirmRedirect() {
		return (String) this.get("redirecturl");
	}

	public long getTokenLife() {
		return Long.parseLong((String)this.get("jwt-lifetime"));
	}

	public long getWebTokenLife() {
		return Long.parseLong((String)this.get("jwt-web-lifetime"));
	}

	public long getRefreshTokenLife() {
		return Long.parseLong((String)this.get("jwt-refresh-lifetime"));
	}

	public long getResetTokenLife() {
		return Long.parseLong((String)this.get("pwd-reset-lifetime"));
	}

	public String getFBAppID() {
		return (String) this.get("facebook-id");
	}

	public String getFBAppSecret() {
		return (String) this.get("facebook-secret");
	}

	public String getAPIKey() {
		return (String) this.get("apikey");
	}

	public String getJWTAudience() {
		return (String) this.get("jwt-audience");
	}

	public String getRefreshAudience() {
		return (String) this.get("jwt-refresh-audience");
	}

	public String getDefaultPassword() {
		return (String) this.get("sas-passwd");
	}


	///map stuff
	@Override
	public int size() {
		return props.size();
	}

	@Override
	public boolean isEmpty() {
		return props.isEmpty();
	}

	@Override
	public boolean containsKey(Object key) {
		String k = (String) key;
		String ctxKey = context + "." + k;
		return props.containsKey(ctxKey) || props.containsKey(k);
	}

	@Override
	public boolean containsValue(Object value) {
		return props.containsValue(value);
	}

	@Override
	public Object get(Object key) {
		String k = (String) key;
		String ctxKey = context + "." + k;
		String result;

		if (props.containsKey(ctxKey)) {
			result = (String) props.get(ctxKey);
		} else {

			result = (String) props.get(key);
		}
		try {
			return groovy.eval(result);
		} catch (Exception ex) {
			return result;
		}
	}

	public boolean getBool(String key) {
		String v = ((String) this.get(key)).toLowerCase();
		if (v == "true" || v == "1" || v == "on") {
			return true;
		} else {
			return false;
		}
	}

	@Override
	public Object put(Object key, Object value) {
		throw new UnsupportedOperationException("Configuration objects are read-only.");
	}

	@Override
	public Object remove(Object key) {
		throw new UnsupportedOperationException("Configuration objects are read-only.");
	}

	@Override
	public void putAll(Map m) {
		throw new UnsupportedOperationException("Configuration objects are read-only.");
	}

	@Override
	public void clear() {
		throw new UnsupportedOperationException("Configuration objects are read-only.");
	}

	@Override
	public Set<?> keySet() {
		return props.keySet();
	}

	@Override
	public Collection<?> values() {
		return props.values();
	}

	@Override
	public Set<?> entrySet() {
		return props.entrySet();
	}

	@Override
	public String toString() {
		StringBuilder result = new StringBuilder();
		result.append("Configuration:\n");
		for (String key : (Set<String>) this.keySet()) {
			Object value = this.get(key);
			result.append("  ");
			result.append(key);
			result.append(" = (");
			result.append(value.getClass().getName());
			result.append(") ");
			result.append(value.toString());
			result.append("\n");
		}
		return result.toString();
	}
}

